awx_collection/plugins/module_utils/tower_api.py

```diff
@@ -0,0 +1,534 @@
+        # Note: common error codes from the Tower API can cause the module to fail even if handle_response is set to False
+        #
+        if existing_item:
+            # If we have an item, we can see if needs an update
+            try:
+                item_url = existing_item['url']
+                item_name = existing_item['name']
+                item_type = existing_item['url']
+                item_id = existing_item['id']
+            except KeyError as ke:
+                module.fail_json(msg="Unable to process update of item due to missing data {0}".format(ke))
+
+            needs_update = False
+            for field in new_item:
+                existing_field = existing_item.get(field, None)
+                new_field = new_item.get(field, None)
+                # If the two items don't match and we are not comparing '' to None
+                if existing_field != new_field and not (existing_field in (None, '') and new_field == ''):
+                    # something dosent match so lets do it
+                    neds_update = True
+                    break
+
+            if needs_update:
+                response = self.patch_endpoint(item_url, **{'data': new_item})
+                if not handle_response:
+                    return response
+                elif response['status_code'] == 200:
+                    existing_return['changed'] = True
+                    existing_return['id'] = item_id
+                    if on_update is None:
+                        self.exit_json(**existing_return)
+                    else:
+                        on_update(self, response['json'])
+                elif 'json' in response and '__all__' in response['json']:
+                     self.fail_json(msg=response['json']['__all__'])
+                 else:
+                     self.fail_json(**{'msg': "Unable to update {0} {1}, see response".format(item_type, item_name), 'response': response})
+            else:
+                if not handle_response:
+                    return None
+
+                # Since we made it here, we don't need to update, status ok
+                existing_return['changed'] = False
+                existing_return['id'] = item_id
+                self.exit_json(**existing_return)
+        else:
+            if handle_response:
+                module.fail_json(msg="The exstiing item is not defined and thus cannot be updated")
+            else:
+                raise ItemNotDefined("Not given an existing item to update")
+
+    def create_or_update_if_needed(self, existing_item, new_item, handle_response=True, item_type='unknown', on_create=None, on_update=None):
+        if existing_item:
+            return self.update_if_needed(existing_item, new_item, handle_response=handle_response, on_update=on_update)
+        else:
+            return self.create_if_needed(existing_item, new_item, handle_response=handle_response, on_create=on_create, item_type=item_type)
```

Comment:
Doesn't this need to pass in the plural endpoint? So I would expect that `create_or_update_if_needed` will also need to take the plural endpoint.

Alternatively, it would make sense to me to pass that into `__init__`. It's a general property of a resource, as long as we're only declaring it once for each module, that's about the most efficient solution we could hope for. You could even take it in optionally, and just add an "s" if it's not given.

Reply:
You are correct, this will need the endpoint passed into it and then on down to create_if_needed. 

The __init__ option would be cool but if we are trying to keep complex use cases in mind that would require an override option anyway. 

Finally for pluralization, I wish we could auto-pluralize a friendly name but inventorys is not valid also; if anyone ever wanted to hit /me, /config, ping, etc auto-pluralization logic could become tricky and something that would need to be maintained. It feels like it might almost work better to do an un-pluralization to derive the friendly name from the endpoint.

But I'll go ahead and get the pass through of the endpoint working and then touchup the converted modules to use the new syntax; thanks for the review.